### DISCUSSION TITLE
Discussion on Changes to netkvm RX Capacity Default Value

### DISCUSSION THREAD

**Discussion by zjmletang** [virtio-win/kvm-guest-drivers-windows#1226] [General]
I’ve noticed that the default value for netkvm RX capacity has been changed from 256 to 1024. The reason provided is to make the configuration more user-friendly, as seen in [RHEL-19430](https://issues.redhat.com/browse/RHEL-19430). My understanding is that this effectively hands over the decision of the default value to the backend. I would like to discuss a few points regarding this change:

1. This change may introduce an issue. If a cloud vendor believes that the most suitable default value for Windows is still 256 (which is quite likely, considering that netkvm still consumes a significant amount of memory), it seems that the only option would be to pass in 256 during the instance creation in QEMU. Once the instance is created, users would not be able to increase the RX capacity beyond 256 within the guest due to the limitation posed by the ring size of 256.

2. Additionally, there might be a minor side effect for cloud vendors who may prefer not to differentiate between Windows and Linux in QEMU. They might choose a unified larger value (such as 1024 or 4096). However, for a larger ring size like 1024, using it directly on Windows could lead to excessive memory usage, making it less suitable as a default value.

3. Lastly, regardless of the backend configuration options or adjustments within the guest, what would be the most appropriate value for the current netkvm? Personally, I still believe that 256 is the optimal choice.

**Comment by ybendito** [comment]
> 1. This change may introduce an issue. If a cloud vendor believes that the most suitable default value for Windows is still 256 (which is quite likely, considering that netkvm still consumes a significant amount of memory), it seems that the only option would be to pass in 256 during the instance creation in QEMU. Once the instance is created, users would not be able to increase the RX capacity beyond 256 within the guest due to the limitation posed by the ring size of 256.

This is exactly the intention of the change

> 2. Additionally, there might be a minor side effect for cloud vendors who may prefer not to differentiate between Windows and Linux in QEMU. They might choose a unified larger value (such as 1024 or 4096). However, for a larger ring size like 1024, using it directly on Windows could lead to excessive memory usage, making it less suitable as a default value.

In this case the cloud vendor can reconfigure the driver to use lower value. As far as I remember, the spec recommends to initialize all the entries in the RX queue.

> 3. Lastly, regardless of the backend configuration options or adjustments within the guest, what would be the most appropriate value for the current netkvm? Personally, I still believe that 256 is the optimal choice.

It is what we had from the beginning and there is another problem - need to change queue size in 2 places to increase it.

**Comment by zjmletang** [comment]
thanks. I would like to ask again whether the RX capacity and TX capacity parameters will be removed from the subsequent inf file

**Reply by ybendito** [reply]
@zjmletang No, we do not have any plans to remove these parameters. BTW, what is the advantage of having RX queue in hardware as large as 4096?

**Reply by zjmletang** [reply]
@ybendito Having an RX queue as large as 4096 in hardware can significantly enhance the handling capability of network interfaces under high traffic loads. In our usage scenarios, the throughput demand for vports often exceeds 10Gbps. Thus, increasing the RX queue size helps improve packet processing efficiency and reduces packet loss, ensuring network performance remains stable and reliable under high bandwidth demands
